cpp/src/reductions/minmax.cu

```diff
@@ -0,0 +1,200 @@
+    : min_val(min_val_), max_val(max_val_), min_valid(min_valid_), max_valid(max_valid_){};
+};
+
+/**
+ * @brief functor that accepts two minmax_pairs and returns a
+ * minmax_pair whose minimum and maximum values are the min() and max()
+ * respectively of the minimums and maximums of the input pairs. Respects
+ * validity.
+ *
+ */
+template <typename T, bool has_nulls = true>
+struct minmax_binary_op
+  : public thrust::binary_function<minmax_pair<T>, minmax_pair<T>, minmax_pair<T>> {
+  __device__ minmax_pair<T> operator()(const minmax_pair<T> &x, const minmax_pair<T> &y) const
+  {
+    T x_min = x.min_valid || !has_nulls ? x.min_val : cudf::DeviceMin::identity<T>();
+    T y_min = y.min_valid || !has_nulls ? y.min_val : cudf::DeviceMin::identity<T>();
+    T x_max = x.max_valid || !has_nulls ? x.max_val : cudf::DeviceMax::identity<T>();
+    T y_max = y.max_valid || !has_nulls ? y.max_val : cudf::DeviceMax::identity<T>();
+
+    // The only invalid situation is if we compare two invalid values.
+    // Otherwise, we are certain to select a valid value due to the
+    // identity functions above changing the comparison value.
+    bool valid_min_result = !has_nulls || x.min_valid || y.min_valid;
+    bool valid_max_result = !has_nulls || x.max_valid || y.max_valid;
+
+    return minmax_pair<T>{
+      thrust::min(x_min, y_min), valid_min_result, thrust::max(x_max, y_max), valid_max_result};
+  }
+};
+
+/**
+ * @brief functor that calls thrust::transform_reduce to produce a std::pair
+ * of scalars that represent the minimum and maximum values of the input data
+ * respectively. Note that dictionaries and non-relationally comparable objects
+ * are not supported.
+ *
+ */
+struct minmax_functor {
+  template <typename T>
+  std::pair<std::unique_ptr<scalar>, std::unique_ptr<scalar>> operator()(
+    const cudf::column_view &col, rmm::mr::device_memory_resource *mr, cudaStream_t stream)
+  {
+    auto device_col = column_device_view::create(col, stream);
+
+    // compute minimum and maximum values
+    minmax_pair<T> result;
+    if (col.nullable()) {
+      result = thrust::transform_reduce(
+        thrust::make_counting_iterator<size_type>(0),
+        thrust::make_counting_iterator<size_type>(col.size()),
+        [d_col = *device_col] __device__(size_type index) -> minmax_pair<T> {
+          return minmax_pair<T>(d_col.element<T>(index), d_col.is_valid(index));
+        },
+        minmax_pair<T>{},
+        minmax_binary_op<T, true>{});
+    } else {
+      result = thrust::transform_reduce(
+        thrust::make_counting_iterator<size_type>(0),
+        thrust::make_counting_iterator<size_type>(col.size()),
+        [d_col = *device_col] __device__(size_type index) -> minmax_pair<T> {
+          return minmax_pair<T>(d_col.element<T>(index), d_col.is_valid(index));
+        },
+        minmax_pair<T>{},
+        minmax_binary_op<T, false>{});
+    }
```

Comment:
Perfect usecase for an [IIFE](https://www.bfilipek.com/2016/11/iife-for-complex-initialization.html)
```suggestion
    minmax_pair<T> const result = [&](){
    auto begin = thrust::make_counting_iterator<size_type>(0);
    auto end = begin + col.size();
    auto op = [d_col = *device_col] __device__(size_type index) -> minmax_pair<T> {
          return minmax_pair<T>(d_col.element<T>(index), d_col.is_valid(index));
        };
     
    if (col.nullable()) {
      return thrust::transform_reduce(begin, end,
        op,
        minmax_pair<T>{},
        minmax_binary_op<T, true>{});
    } else {
      return thrust::transform_reduce(begin, end, op       
        minmax_pair<T>{},
        minmax_binary_op<T, false>{});
    }();
```

Reply:
I originally tried to do this, but I was focused on the only change, the binary op. At the time, I had two different classes derived from the same parent class, but as they functions were not virtual, I was unable to figure out a way to assign a variable to the correct op. I was trying something like `auto op = [])() { if (nullable) { return op1; } else { return op2; } }();`. It didn't occur to me to return the entire thing. Thanks!

Reply:
Yeah, it is unfortunate that you can't just conditionally store one of the two ops into an `auto` variable and use that. It doesn't work since they are two different types.